src/full/Agda/TypeChecking/Sort.hs

```diff
@@ -85,57 +86,66 @@ hasBiggerSort = void . inferUnivSort
 -- | Infer the sort of a pi type. If we can compute the sort straight away,
 --   return that. Otherwise, return @PiSort a s2@ and add a constraint to
 --   ensure we can compute the sort eventually.
-inferPiSort :: PureTCM m => Dom Type -> Abs Sort -> m Sort
+inferPiSort :: (PureTCM m,MonadConstraint m) => Dom Type -> Abs Sort -> m Sort
 inferPiSort a s2 = do
-  s1' <- reduce $ getSort a
-  s2' <- mapAbstraction a reduce s2
-  -- we do instantiateFull here to perhaps remove some (flexible)
-  -- dependencies of s2 on var 0, thus allowing piSort' to reduce
-  s2' <- instantiateFull s2'
-
+  s1' <- reduceB $ getSort a
+  s2' <- mapAbstraction a reduceB s2
   --Jesper, 2018-04-23: disabled PTS constraints for now,
   --this assumes that piSort can only be blocked by unsolved metas.
+  --Arthur Adjedj, 2023-02-27, Turned PTS back on,
+  --piSort can now be blocked by Leveluniv
+  let b' = unblockOnEither (getBlocker s1') (getBlocker $ unAbs s2')
+  case piSort' (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2') of
+    Right s -> return s
+    Left b -> do
+      addConstraint (unblockOnEither b b') $ HasPTSRule a (ignoreBlocking <$> s2')
+      return $ PiSort (unEl <$> a) (ignoreBlocking s1') (ignoreBlocking <$> s2')
 
-  --case piSort' s1 s2 of
-  --  Just s -> return s
-  --  Nothing -> do
-  --    addConstraint $ HasPTSRule s1 s2
-  --    return $ PiSort s1 s2
-
-  return $ piSort (unEl <$> a) s1' s2'
 
 -- | As @inferPiSort@, but for a nondependent function type.
 inferFunSort :: Sort -> Sort -> TCM Sort
-inferFunSort s1 s2 = funSort <$> reduce s1 <*> reduce s2
-
-ptsRule :: Dom Type -> Abs Sort -> Sort -> TCM ()
-ptsRule a b c = do
-  c' <- inferPiSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
-
--- | Non-dependent version of ptsRule
-ptsRule' :: Sort -> Sort -> Sort -> TCM ()
-ptsRule' a b c = do
-  c' <- inferFunSort a b
-  ifM (optCumulativity <$> pragmaOptions)
-    (leqSort c' c)
-    (equalSort c' c)
+inferFunSort s1 s2 = do
+  ss1 <- sortOf (Sort s1)
+  s1' <- reduceB s1
+  s2' <- reduceB s2
+  case funSort' (ignoreBlocking s1') (ignoreBlocking s2') of
+    Right s -> return s
+    Left b -> do
+      let b' = unblockOnEither (getBlocker s1') (getBlocker s2')
+      addConstraint (unblockOnEither b b') $ HasPTSRule (defaultDom (El (univSort s1) (Sort s1))) (mkAbs "_" $ ignoreBlocking s2')
+      return $ FunSort s1 s2
```

Comment:
@arthur-adjedj : Why changing `NoAbs` (which has no computational costs) for `mkAbs` (which has to check whether the 0th variable occurs in `s2'`)?  Is it now the case that we may need to create a `PiSort` even though we started out with a `FunSort`?  I would suppose not, but if this is the case, it deserves a comment.
CC @jespercockx

Reply:
The `mkAbs` here actually seems *wrong* to me as it adds a non-existing variable to the scope of the codomain sort. `NoAbs` would be the right thing to use here. Even better would be to have a non-dependent version of the `HasPTSRule` constraint so these kind of mistakes could also be prevented in the future.